# Basic Data Manipulation

* Once your data has been loaded as a Dataframe, you can start using Pandas various functions to quickly explore your data 

In [2]:
# load pandas
import pandas as pd

# load data
center_attendance_pandas = pd.read_csv("community-center-attendance.csv", 
                                       index_col="_id") # use the column named _id as the row index

### Helpful functions for exploring Dataframes and Series

* Looking at parts of the Dataframe
* `<dataframe>.head(n)` - look at the first n rows of the dataframe
* `<dataframe>.tail(n)` - look at the last n rows of the dataframe
* `<dataframe>.sample(n)` - randomly select n rows from the dataframe

In [3]:
# Look at the first 10 rows
center_attendance_pandas.head(10)

date                  center_name  attendance_count
_id                                                           
1    2018-06-08      Ormsby Community Center                10
2    2018-06-08     Paulson Community Center                19
3    2018-06-08    Phillips Community Center               107
4    2018-06-08       Ammon Community Center                81
5    2018-06-08   Brookline Community Center                33
6    2018-06-08   Jefferson Community Center                29
7    2018-06-08  Warrington Community Center                15
8    2018-06-08   West Penn Community Center                54
9    2018-06-07    Phillips Community Center                77
10   2018-06-07     Paulson Community Center                25

In [4]:
# Look at the last 5 rows
center_attendance_pandas.tail()

date                  center_name  attendance_count
_id                                                             
18363  2011-03-08       Magee Community Center                32
18364  2011-03-08   West Penn Community Center                 3
18365  2011-03-07  Warrington Community Center                 1
18366  2011-03-07       Magee Community Center                 7
18367  2011-03-07   West Penn Community Center                 2

In [5]:
# Grab 5 random rows
center_attendance_pandas.sample(5)

date                 center_name  attendance_count
_id                                                            
329    2018-04-13   Phillips Community Center               107
10916  2014-02-14  Arlington Community Center                48
11609  2013-11-07  Jefferson Community Center                10
10884  2014-02-19  Brookline Community Center               382
2818   2017-03-27  Brookline Community Center                82

* How many rows and columns?
* `<dataframe>.shape` - return the rows and columns as a python data structure (not a function!)
* `<dataframe>.info()` - Display the datatypes of the index and columns as well as memory usage
* `<dataframe>.describe()` - Compute summary statistics for numerical columns

In [6]:
# How many rows and columns
center_attendance_pandas.shape

(18367, 3)

In [7]:
# Inspect the datatypes
center_attendance_pandas.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18367 entries, 1 to 18367
Data columns (total 3 columns):
date                18367 non-null object
center_name         18367 non-null object
attendance_count    18367 non-null int64
dtypes: int64(1), object(2)
memory usage: 574.0+ KB


* The output above shows us a lot of implementation details about our dataframe
* Data types, number of rows and columns, and the datatype of the column
* Also shows us memory usage, which is useful because memory is a limited resource

* We can also start doing some computations on the data

In [8]:
# Compute summary statistics on the numerical columns
center_attendance_pandas.describe()

attendance_count
count      18367.000000
mean          61.934883
std           58.510201
min            1.000000
25%           24.000000
50%           43.000000
75%           80.000000
max          741.000000

* The `describe()` function will automatically compute summary statistics for numerical columns and ignore categorical columns

### Counting Numerical Data

* We can use traditional Python functions to get information about our Dataframe.
* The `len()` function tells us the length of the sequence 

In [9]:
# use a standard python function to get the length of the sequence
len(center_attendance_pandas)

18367

* So this tells us our dataset has 18,367 rows.
* But this is just information about the dataset itself, it doesn't tell us how many people visited community centers
* How many people visited all the community centers for all time (in the dataset)?
* First let's answer this using pure python

In [10]:
# Load up the CSV module and center attendance in python data structures
import csv

with open('community-center-attendance.csv') as f:
    center_attendance_python = [row for row in csv.reader(f)]

# look at the first ten rows of the data loaded in python
center_attendance_python[0:10]

[['_id', 'date', 'center_name', 'attendance_count'],
 ['1', '2018-06-08', 'Ormsby Community Center', '10'],
 ['2', '2018-06-08', 'Paulson Community Center', '19'],
 ['3', '2018-06-08', 'Phillips Community Center', '107'],
 ['4', '2018-06-08', 'Ammon Community Center', '81'],
 ['5', '2018-06-08', 'Brookline Community Center', '33'],
 ['6', '2018-06-08', 'Jefferson Community Center', '29'],
 ['7', '2018-06-08', 'Warrington Community Center', '15'],
 ['8', '2018-06-08', 'West Penn Community Center', '54'],
 ['9', '2018-06-07', 'Phillips Community Center', '77']]

In [11]:
# create a variable to hold the total attendance
total_attendance = 0

# loop over the data that was loaded using pure python
for row in center_attendance_python[1:]: # skip the header row using a list slice
    # add the row count to the total
    row_attendance = row[3]
    total_attendance = total_attendance + row_attendance

print(total_attendance)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

* oh shoot! The csv module always opens everything as a string
* We need to add some type coercion

In [12]:
# create a variable to hold the total attendance
total_attendance = 0

# loop over the data that was loaded using pure python
for row in center_attendance_python[1:]: # skip the header row using a list slice
    # add the row count to the total, convert string to int
    row_attendance = int(row[3])
    total_attendance = total_attendance + row_attendance

print(total_attendance)

1137558


* Now here is how we do the exact same thing with Pandas 😄
* This code selects the `attendance_count` column and then computes the sum of all the values.

In [13]:
# compute the total attendance with the pandas sum function
center_attendance_pandas['attendance_count'].sum()

1137558

### Summary Statistics

* We can also look at the summary statistics individually
* `<dataframe>[<column name>].mean()` - calculate the mean value for the column values
* `<dataframe>[<column name>].std()` - calculate the standard deviation for the column values
* `<dataframe>[<column name>].var()` - calculate the variance value for the column values
* `<dataframe>[<column name>].median()` - calculate the median value for the column values
* `<dataframe>[<column name>].min()` - calculate the minimum value for the column values

In [14]:
# mean attendance per day at all community centers
center_attendance_pandas['attendance_count'].mean()

61.93488321446072

In [15]:
# standard deviation
center_attendance_pandas['attendance_count'].std()

58.51020126797415

In [16]:
# variance
center_attendance_pandas['attendance_count'].var()

3423.4436524188445

In [17]:
# median attendance per day at all community centers
center_attendance_pandas['attendance_count'].median()

43.0

In [18]:
# minimum attendance at community centers
center_attendance_pandas['attendance_count'].min()

1

* Note, missing values are automatically skipped unless the entire column is NA.

In [19]:
df = pd.DataFrame([[1.4, None], [7.1, -4.5], 
                  [None, None], [0.75, -1.3]],
                 index=['a','b','c','d'],
                 columns=['one','two'])
df

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [20]:
df.sum()

one    9.25
two   -5.80
dtype: float64

In [21]:
df.sum(axis=1)

a    1.40
b    2.60
c    0.00
d   -0.55
dtype: float64

In [22]:
df.mean(axis=1)

a    1.400
b    1.300
c      NaN
d   -0.275
dtype: float64

In [23]:
df.mean(axis=1, skipna=False)

a      NaN
b    1.300
c      NaN
d   -0.275
dtype: float64

### Math Operations

* You can do mathematical operations that will get applied to ever value in the row or column

In [24]:
# create a numerical dataframe
df = pd.DataFrame([[1.4, 4.7], [7.1, -4.5], 
                  [3, 7], [0.75, -1.3]],
                 index=['a','b','c','d'],
                 columns=['one','two'])
df

one  two
a  1.40  4.7
b  7.10 -4.5
c  3.00  7.0
d  0.75 -1.3

In [25]:
# multiple two columns against eachother
df['one'] * df['two']

a     6.580
b   -31.950
c    21.000
d    -0.975
dtype: float64

In [26]:
# divide a column by a number
df['one'] / 5

a    0.28
b    1.42
c    0.60
d    0.15
Name: one, dtype: float64

* Pandas is not only a tool for working with numerical data
* Lots of functionality for manipulating categorical data too

### Counting Categorical Data

* Just like before we can start counting the distribution of values in the column. 
* how many entries per community center (this isn't counting attendance but counting the number of rows per center).

* The "Pythonic way"

In [27]:
# Create a dictionary to store the counts
center_counter = dict()

# loop over the data
for row in center_attendance_python[1:]:
    center = row[2]
    
    # check to see if the gender is already in the diction
    if center not in center_counter:
        # create a new entry
        center_counter[center] = 1
    else:
        # increment a new entry
        #center_counter[center] += 1
        center_counter[center] = center_counter[center] + 1

# Display the dictionary 
center_counter

{'Ammon / Josh Gibson Field': 1,
 'Ammon Community Center': 1825,
 'Ammon Pool': 3,
 'Arlington Community Center': 1331,
 'Arlington Field (Playground)': 1,
 'Brookline Community Center': 2159,
 'Dan Marino Field (Playground)': 1,
 'Frick Environmental Center': 1,
 'Gladstone Field': 5,
 'Highland Pool': 1,
 'Jefferson Community Center': 1701,
 'Magee Community Center': 1800,
 'Mellon Tennis Center': 6,
 'Moore Pool': 1,
 'Ormsby Community Center': 1990,
 'Ormsby Field (Playground)': 8,
 'Paulson Community Center': 1547,
 'Paulson Field': 3,
 'Phillips Community Center': 2116,
 'Phillips Park Field': 13,
 'Schenley Ice Rink': 1,
 'Warrington Community Center': 1714,
 'Warrington Field': 1,
 'West Penn Community Center': 2130,
 'West Penn Fields': 1,
 'West Penn Pool': 7}

* The Pandas way is a bit easier

In [28]:
# Do the same thing with pandas
center_attendance_pandas['center_name'].value_counts()

Brookline Community Center       2159
West Penn Community Center       2130
Phillips Community Center        2116
Ormsby Community Center          1990
Ammon Community Center           1825
Magee Community Center           1800
Warrington Community Center      1714
Jefferson Community Center       1701
Paulson Community Center         1547
Arlington Community Center       1331
Phillips Park Field                13
Ormsby Field (Playground)           8
West Penn Pool                      7
Mellon Tennis Center                6
Gladstone Field                     5
Ammon Pool                          3
Paulson Field                       3
Schenley Ice Rink                   1
Arlington Field (Playground)        1
Moore Pool                          1
Dan Marino Field (Playground)       1
Ammon / Josh Gibson Field           1
West Penn Fields                    1
Frick Environmental Center          1
Warrington Field                    1
Highland Pool                       1
Name: center

In [29]:
# change the sort order
center_attendance_pandas['center_name'].value_counts().sort_values(ascending=True)

Highland Pool                       1
Frick Environmental Center          1
West Penn Fields                    1
Ammon / Josh Gibson Field           1
Dan Marino Field (Playground)       1
Moore Pool                          1
Arlington Field (Playground)        1
Schenley Ice Rink                   1
Warrington Field                    1
Paulson Field                       3
Ammon Pool                          3
Gladstone Field                     5
Mellon Tennis Center                6
West Penn Pool                      7
Ormsby Field (Playground)           8
Phillips Park Field                13
Arlington Community Center       1331
Paulson Community Center         1547
Jefferson Community Center       1701
Warrington Community Center      1714
Magee Community Center           1800
Ammon Community Center           1825
Ormsby Community Center          1990
Phillips Community Center        2116
West Penn Community Center       2130
Brookline Community Center       2159
Name: center

## A Few Other Aggregation & Statistical Methods

* There are a bunch of other DataFrame and Series methods for working with data

In [30]:
# select unique values from column, returns a numpy array
center_attendance_pandas['center_name'].unique()

array(['Ormsby Community Center', 'Paulson Community Center',
       'Phillips Community Center', 'Ammon Community Center',
       'Brookline Community Center', 'Jefferson Community Center',
       'Warrington Community Center', 'West Penn Community Center',
       'Magee Community Center', 'Arlington Community Center',
       'Gladstone Field', 'Mellon Tennis Center', 'Phillips Park Field',
       'Schenley Ice Rink', 'Warrington Field', 'West Penn Fields',
       'West Penn Pool', 'Dan Marino Field (Playground)', 'Paulson Field',
       'Ormsby Field (Playground)', 'Ammon Pool', 'Moore Pool',
       'Frick Environmental Center', 'Arlington Field (Playground)',
       'Highland Pool', 'Ammon / Josh Gibson Field'], dtype=object)

In [31]:
df = pd.DataFrame([[1.4, 10], [7.1, -4.5], 
                  [2.6, 0], [0.75, -1.3]],
                 index=['a','b','c','d'],
                 columns=['one','two'])
df

one   two
a  1.40  10.0
b  7.10  -4.5
c  2.60   0.0
d  0.75  -1.3

In [32]:
# Does the column contain any zero, empty or false
df.all()

one     True
two    False
dtype: bool

In [33]:
# Return true if there are any elements in the column
df.any()

one    True
two    True
dtype: bool

In [34]:
# Count the number of entries
center_attendance_pandas.count()

date                18367
center_name         18367
attendance_count    18367
dtype: int64

In [35]:
# What is the max value in a specified column
center_attendance_pandas['attendance_count'].max()

741

In [36]:
# What is the index of the max value in specified column
center_attendance_pandas['attendance_count'].idxmax()

14707

In [37]:
# use the output of idxmax() to select the whole row
center_attendance_pandas.iloc[center_attendance_pandas['attendance_count'].idxmax()]

date                                2012-09-18
center_name         Brookline Community Center
attendance_count                           741
Name: 14707, dtype: object

---

## Subsetting Data

* It is sometimes helpful to think of a Pandas Dataframe as a little database. 
* There is data and information stored in the Pandas Dataframe (or Series) and you want to *retrieve* it.
* Pandas has multiple mechanisms for getting specific bits of data and information from its data structures. 

### Masking: Filtering by Values

* The most common is to use *masking* to select just the rows you want. 
* Masking is a two stage process, first you create a sequence of boolean values based upon a conditional expression--which you can think of as a "query"--and then you index your dataframe using that boolean sequence. 

In [38]:
# read the data into a pandas dataframe, using the "_id" column for index
order_data  = pd.read_csv("chipotle.tsv", sep="\t")
# inspect the dataframe
order_data.head() 

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

In [39]:
# Let's look at the chipotle order data
order_data.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
6                                                NaN        1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...        9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        9.25

In [40]:
# Let's look at all the columns
order_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
order_id              4622 non-null int64
quantity              4622 non-null int64
item_name             4622 non-null object
choice_description    3376 non-null object
item_price            4622 non-null float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.6+ KB


* How might we only look at particular orders?
* First step is to create a *query mask*, a list of `True/False` values for rows that satisfy a particular condition.

In [41]:
# create a query mask for chicken bowls
query_mask = order_data['item_name'] == "Chicken Bowl"

#look at the first 20 items to see what matches
query_mask.head(20)

0     False
1     False
2     False
3     False
4      True
5      True
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13     True
14    False
15    False
16    False
17    False
18    False
19     True
Name: item_name, dtype: bool

* This tells us the row id and True or False if the item type equals chicken bowl
* We can look up that row by index and see if it is correct

In [42]:
order_data.iloc[19]

order_id                                                             10
quantity                                                              1
item_name                                                  Chicken Bowl
choice_description    [Tomatillo Red Chili Salsa, [Fajita Vegetables...
item_price                                                         8.75
Name: 19, dtype: object

* Yup! So now that we know the mask works, we can create a *subset* of our data containing chicken bowls.

In [43]:
chicken_bowls = order_data[query_mask]
chicken_bowls.head()

order_id  quantity     item_name  \
4          2         2  Chicken Bowl   
5          3         1  Chicken Bowl   
13         7         1  Chicken Bowl   
19        10         1  Chicken Bowl   
26        13         1  Chicken Bowl   

                                   choice_description  item_price  
4   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
13  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...        8.75  
26  [Roasted Chili Corn Salsa (Medium), [Pinto Bea...        8.49

* Now you can do things like calculate the average price for chicken bowl orders

In [44]:
# Calculate the mean price for chicken bowls
chicken_bowls['item_price'].mean()

10.113953168044079

In [45]:
# See how many chicken bowls people order
chicken_bowls['quantity'].value_counts()

1    693
2     31
3      2
Name: quantity, dtype: int64

* We can also combine query masks using boolean logic
* Can we look at just the chicken bowl orders that were less than $10

In [46]:
# create a query mask for chicken bowls
item_query_mask = order_data['item_name'] == "Chicken Bowl"
# create a query mask for cheap orders
price_query_mask = order_data['item_price'] < 10

# apply both query masks using boolean AND
cheap_chicken_bowls = order_data[item_query_mask & price_query_mask]
cheap_chicken_bowls.head()

order_id  quantity     item_name  \
19        10         1  Chicken Bowl   
26        13         1  Chicken Bowl   
76        34         1  Chicken Bowl   
78        34         1  Chicken Bowl   
99        44         1  Chicken Bowl   

                                   choice_description  item_price  
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...        8.75  
26  [Roasted Chili Corn Salsa (Medium), [Pinto Bea...        8.49  
76  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        8.75  
78  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...        8.75  
99  [Tomatillo Red Chili Salsa, [Rice, Fajita Vege...        8.75

In [47]:
# Median price for cheap chicken bowls
cheap_chicken_bowls['item_price'].median()

8.75

* Query masks can be used to filter and create subsets of data
* Note, this method of subsetting data creates what is called a "view" of the data
* You are basically working with a big slice of the original dataframe, not a separate copy of the data
* This means if you try an do transformations on that view, you will get an error
* For more information, [see the pandas documentation](http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy)

In [48]:
cheap_chicken_bowls['half_price'] = cheap_chicken_bowls['item_price'] / 2

/home/mcburton/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [49]:
copy_of_cheap_chicken_bowls = cheap_chicken_bowls.copy()
copy_of_cheap_chicken_bowls['half_price'] = copy_of_cheap_chicken_bowls['item_price'] / 2
copy_of_cheap_chicken_bowls.head()

order_id  quantity     item_name  \
19        10         1  Chicken Bowl   
26        13         1  Chicken Bowl   
76        34         1  Chicken Bowl   
78        34         1  Chicken Bowl   
99        44         1  Chicken Bowl   

                                   choice_description  item_price  half_price  
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...        8.75       4.375  
26  [Roasted Chili Corn Salsa (Medium), [Pinto Bea...        8.49       4.245  
76  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        8.75       4.375  
78  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...        8.75       4.375  
99  [Tomatillo Red Chili Salsa, [Rice, Fajita Vege...        8.75       4.375